### ISSUE TITLE
[Debug] Launch In Terminal still goes to the previous terminal window after i change the default shell

### ISSUE THREAD

**Issue by testforstephen** [microsoft/vscode#81345]
- VSCode Version: 
```
Version: 1.39.0-insider (system setup)
Commit: de03947834d0bcf55093b97c292b76faa75cdebc
Date: 2019-09-23T08:25:10.478Z
Electron: 6.0.9
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.18362
```
- OS Version:  Windows 10

Steps to Reproduce:
1. Create HelloWorld.java in VS Code.
2. Set cmd as the default shell.
3. Click Run codelens on top of main method, it will open a cmd terminal to run the program.
![image](https://user-images.githubusercontent.com/14052197/65474472-8f380600-deae-11e9-94f4-ff252e02b04d.png)
4. Change the default shell to powershell, and click Run codelens again.
5. The generated command line style is powershell style, but VS Code still run it in the previous cmd terminal so that it failed. But if i kill all existing terminals, clicking Run codelens will be able to launch the correct terminal.
![image](https://user-images.githubusercontent.com/14052197/65474858-fb673980-deaf-11e9-833e-c721fb5adc7e.png)

Expected: Launch In Terminal should open the correct terminal to run the program.

**Comment by Tyriar** [comment]
Whatever extension is creating this terminal is choosing which shell to use:

![image](https://user-images.githubusercontent.com/2193314/65634506-368e6900-df93-11e9-8c04-c5a342765fbe.png)

Also it sounds like this codelens thing is sending text to the current terminal, that sounds like a bug in that extension.

**Comment by testforstephen** [comment]
I have two questions here:
- How did VS Code determine whether to reuse the current terminal? By the terminal title?
- As a debugger, we don't have knowledge about whether the shell is changed or not. We just follow the [runInTerminal](https://microsoft.github.io/debug-adapter-protocol/specification#Reverse_Requests_RunInTerminal) request in DAP protocol to implement the launch in terminal. How did the debugger fix this issue?

**Comment by Tyriar** [comment]
@weinand debug doesn't respecting the default shell? By design?

**Comment by int19h** [comment]
We're also seeing this issue in the Python debugger, e.g. https://github.com/microsoft/debugpy/issues/245

And it's not just that "runInTerminal" doesn't respect the shell - it also seems to get confused about which shell it's running in. Suppose I start with my default shell set to PS, and do F5:
```
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Try the new cross-platform PowerShell https://aka.ms/pscore6

PS C:\git\debugpy>  ${env:DEBUGPY_LOG_DIR}='c:\Users\int19\.vscode\extensions\ms-python.python-2020.5.78807'; & 'C:\Python\3.7-64\python.exe' 'c:\Users\int19\.vscode\extensions\ms-python.python-2020.5.78807\pythonFiles\lib\python\debugpy\wheels\debugpy\launcher' '59043' '--' 'c:\Temp\test3.py'
```
So far, so good - the shell is as expected, and the command uses PS syntax to set environment variables. But suppose I stop debugging, change the default to cmd.exe, and immediately do F5 again:
```
PS C:\git\debugpy>  cd c:\git\debugpy && cmd /C "set "DEBUGPY_LOG_DIR=c:\Users\int19\.vscode\extensions\ms-python.python-2020.5.78807" && C:\Python\3.7-64\python.exe c:\Users\int19\.vscode\extensions\ms-python.python-2020.5.78807\pythonFiles\lib\python\debugpy\wheels\debugpy\launcher 60317 -- c:\Temp\test3.py "
At line:1 char:20
+  cd c:\git\debugpy && cmd /C "set "DEBUGPY_LOG_DIR=c:\Users\int19\.vs ...
+                    ~~
The token '&&' is not a valid statement separator in this version.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : InvalidEndOfLine
```
Note that it's still using the same terminal - one running PS! - but the command submitted to it is clearly meant for cmd.exe, which is why it failed. However, the part that caused it to fail is only present if "cwd" was specified in the "runInTerminal" request; without "cwd", it would have actually worked as expected, with PS exec'ing cmd, and cmd then taking care of everything else (so this might actually be hit a lot, and the users hitting it aren't even noticing unless they're paying attention to the terminal).

I wonder if this could be related to https://github.com/microsoft/vscode/issues/82662 - we could never figure out why VSCode sometimes decides to create a new terminal for F5, and sometimes reuses one.

**Comment by int19h** [comment]
Oh, and the workaround is to use Kill Terminal to force a new one being created - the new terminal then respects the changed shell setting.

**Comment by kishankarun** [comment]
I too have the problem mentioned in:

Launch In Terminal (Git Bash) didn't reuse the previous terminal window (#81344)

Looks like #81344 is locked.

Could you please provide some help?

Thanks in advance...

**Comment by huntertran** [comment]
Is there any update on this issue?
The issue #81344 is not a duplicate of this

**Comment by weinand** [comment]
I'm reusing a terminal only if it is not blocked by some process **and** if it was created with the same configuration (shell and arguments).
Please try in the next Insiders.
